reaper.go

```diff
@@ -0,0 +1,188 @@
+// Licensed under the Apache License, Version 2.0
+// Details: https://raw.githubusercontent.com/maniksurtani/quotaservice/master/LICENSE
+
+package quotaservice
+
+import (
+	"github.com/maniksurtani/quotaservice/config"
+	"github.com/maniksurtani/quotaservice/logging"
+	pbconfig "github.com/maniksurtani/quotaservice/protos/config"
+	"sync"
+	"time"
+)
+
+// watcher watches reapableBuckets for activity.
+type watcher struct {
+	ns           string
+	bucketName   string
+	identifier   string
+	maxIdle      time.Duration
+	lastActivity time.Time
+	activities   chan struct{}
+}
+
+// activityDetected tells you if activity has been detected since the last time this method was
+// called.
+func (w *watcher) activityDetected() bool {
+	select {
+	case <-w.activities:
+		return true
+	default:
+		return false
+	}
+}
+
+// tooIdle returns true if a bucket has been idle for longer than its maxIdle
+func (w *watcher) tooIdle(now time.Time) bool {
+	if w.activityDetected() {
+		w.lastActivity = now
+		return false
+	}
+	if now.Sub(w.lastActivity) > w.maxIdle {
+		return true
+	} else {
+		return false
+	}
+}
+
+// reapableBucket is a wrapper around a bucket that overrides ReportActivity(), and reports any
+// activity on a channel that is monitored by the reaper.
+type reapableBucket struct {
+	Bucket
+	activities chan struct{}
+}
+
+// ReportActivity is overridden to report activity to the reapableBucket's activity channel.
+func (r *reapableBucket) ReportActivity() {
+	select {
+	case r.activities <- struct{}{}:
+	// reported activity
+	default:
+		// Already reported
+	}
+
+	// Now call ReportActivity() on the delegate bucket.
+	r.Bucket.ReportActivity()
+}
+
+// Destroy closes the reapableBucket's activity channel.
+func (r *reapableBucket) Destroy() {
+	close(r.activities)
+
+	// Now call Destroy() on the delegate bucket.
+	r.Bucket.Destroy()
+}
+
+func (r *reapableBucket) createWatcher(ns, bucketName string, maxIdle time.Duration) *watcher {
+	return &watcher{
+		ns:         ns,
+		bucketName: bucketName,
+		identifier: config.FullyQualifiedName(ns, bucketName),
+		maxIdle:    maxIdle,
+		activities: r.activities}
+}
+
+type reaper struct {
+	sync.Mutex
+	cfg         config.ReaperConfig
+	newWatchers chan *watcher
+	watchers    map[string]*watcher
+	running     bool
+}
+
+func newReaper(bc *bucketContainer, r config.ReaperConfig) *reaper {
+	reaper := &reaper{
+		cfg:         r,
+		watchers:    make(map[string]*watcher),
+		running:     true,
+		newWatchers: make(chan *watcher, r.BucketWatcherBuffer)}
+
+	go reaper.checkForNewWatchers()
+	go reaper.reapIdleBuckets(bc)
+
+	return reaper
+}
+
+func (r *reaper) stop() {
+	r.Lock()
+	defer r.Unlock()
+
+	close(r.newWatchers)
+	r.running = false
+	r.newWatchers = nil
+	r.watchers = nil
+}
+
+func (r *reaper) checkForNewWatchers() {
+	t := time.NewTicker(r.cfg.BucketWatcherFrequency)
+
+	for range t.C {
+		if !r.checkIsRunning() {
+			break
+		}
+
+		for w := range r.newWatchers {
```

Comment:
Are you worried about the cost of locking and adding a new watcher? That should happen more frequently than the reaper itself, so I don't think it'd hurt too much.